packages/react-reconciler/src/ReactFiberBeginWork.js

```diff
@@ -169,12 +169,16 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
 
   function updateForwardRef(current, workInProgress) {
     const render = workInProgress.type.render;
-    const nextChildren = render(
-      workInProgress.pendingProps,
-      workInProgress.ref,
-    );
+    const nextProps = workInProgress.pendingProps;
+    if (hasLegacyContextChanged()) {
+      // Normally we can bail out on props equality but if context has changed
+      // we don't do the bailout and we have to reuse existing props instead.
+    } else if (workInProgress.memoizedProps === nextProps) {
```

Comment:
I think we should also compare `ref`

Reply:
With what?

Reply:
>maybe Foo gets new properties that causes it to recreate the ref, but only a filtered set of those props are passed through to the ForwardRef component, so from it's POV nothing has changed?

OK, that sounds potentially plausible. If you find any free time to hack on this I'd appreciate a test.

Reply:
I think maybe this is too contrived to be a real concern.

Ref callbacks would be re-invoked if the underlying instance/ref changed. If the ref callback itself changed, most likely this just means it's an arrow functions and has an identical implementation.

`createRef` would be more problematic, but it is only expected to be created once in the first place, so in practice shouldn't be a problem.

Maybe we should just add an inline comment saying this is technically possible but probably not a concern in practice?

Reply:
>If the ref callback itself changed, most likely this just means it's an arrow functions and has an identical implementation.

I don’t think we can rely on something like this. Otherwise we might as well never update refs.

We handled this correctly so far so I don’t think we should regress in this particular case.

Reply:
I tried to write a contrived test to catch the case we're talking about, and it looks like I was mistaken to begin with. Even if props are shallowly equal, the `memoizedProps` and `pendingProps` objects themselves won't be equal- so we won't bail out- which side steps the issue I mentioned.

Sorry for the dead end. 😄

Reply:
We chatted a bit with @acdlite and he proposed to check against `current.ref`. Seems like it wouldn’t hurt but I’ll think about it more tomorrow.